### PR TITLE
Laboratory work #3, Aisina Alina - 21FPL1

### PR DESCRIPTION
**Pull request by itsaisina**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -51,7 +52,10 @@ def _clean_and_tokenize(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens
         """
-        pass
+        for mark in self._punctuation:
+            text = text.replace(mark, '')
+        split_text = text.split(' ')
+        return tuple(word1 for word in split_text if (word1 := word.lower()))
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:58]:
```suggestion
        tuple(word.lower() for word in split_text if word)
```
?

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -385,7 +453,17 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        if vertex1 not in self._edges:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:458]: create a separate function that has 2 arguments and use twice

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
try to do for 10

File: lab_3_keywords_textrank/main.py (hunk 1 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: changed

File: lab_3_keywords_textrank/main.py (hunk 2 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -109,7 +118,7 @@ def __init__(self) -> None:
         """
         Constructs all the necessary attributes for the text encoder object
         """
-        pass
+        self._word2id, self._id2word = {}, {}
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:121]: please, write it in two lines

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -120,7 +129,9 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        for index, token in enumerate(tokens):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:132]: read about the `start` argument of the `enumerate` function

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -136,7 +147,11 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+        self._learn_indices(tokens)
+        tokens_list = [self._word2id[key] for token in tokens for key in self._word2id if token == key]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:153]: why do you create list and than cast it to tuple? Can you create tuple immediately?

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -152,7 +167,13 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        if not encoded_tokens:
+            return None
+        for token in encoded_tokens:
+            if token not in self._id2word.keys():
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:173]:
```suggestion
            if token not in self._id2word:
```

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -152,7 +167,13 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
...
+                return None
+        tokens_list = [self._id2word[key] for token in encoded_tokens for key in self._id2word if token == key]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:175]: again

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -173,7 +194,20 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+    final = []
+    tokens_list = list(tokens)
+    for i in range(len(tokens)):
+        window_tokens = tokens_list[i:window_length + i]
+        for j in window_tokens:
+            for k in window_tokens:
+                if (j, k) in final or (k, j) in final:
+                    continue
+                if not j == k:
+                    pair = tuple([j, k])
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:208]:
```suggestion
                    pair = (j, k)
```

File: lab_3_keywords_textrank/main.py [hunk 9]
```diff
@@ -233,7 +269,24 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vertex in vertex1, vertex2:
+            if vertex not in self._matrix[0]:
+                self._matrix[0].append(vertex)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:276]: do not use `self._matrix[0]` as a vertexes storage. You can create another class property for this

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 8 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 4 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 3 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 9 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 10]
```diff
@@ -136,7 +148,11 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+        self._learn_indices(tokens)
+        tokens_list = tuple(self._word2id[key] for token in tokens for key in self._word2id if token == key)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:154]: Why do you need the second for ?

File: lab_3_keywords_textrank/main.py [hunk 11]
```diff
@@ -251,7 +301,10 @@ def is_incidental(self, vertex1: int, vertex2: int) -> int:
                 1 if vertices are incidental, otherwise 0
         If either of vertices is not present in the graph, -1 is returned
         """
-        pass
+        if vertex1 not in self._vertex_list or vertex2 not in self._vertex_list:
+            return -1
+        equal = self._matrix[self._vertex_list.index(vertex1)][self._vertex_list.index(vertex2)]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:306]: equal is a very strange name

File: lab_3_keywords_textrank/main.py (hunk 10 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 11 above)
Comment (itsaisina) [reply lab_3_keywords_textrank/main.py]: done

**Review by artyomtugaryov — APPROVED** [review]

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**